### PR TITLE
fix(elements): fire custom element output events during component initialization

### LINKED ISSUE #36141: Custom Element output events not firing during initialization lifecycle hooks
# 🐞 bug report

### Affected Package
The issue is caused by package @angular/elements

### Is this a regression?
No

### Description
Event listeners on Custom Element component outputs don't fire during `ngOnChanges` (initial call), `ngOnInit`, `ngAfterViewInit`, or `ngAfterContentInit`.

## 🔬 Minimal Reproduction
https://github.com/antch/angular-elements-lifecycle-hooks
For both the Custom and Native element there is a list which should be appended with the fired events. For the native element, we can see the callback fires during initialization of the component, but this doesn't happen for the Custom Element.

You can look in the console to see the execution of the lifecycle hooks.

The reproduction uses artifacts from a [CI build](https://circleci.com/gh/angular/angular/657041#artifacts/containers/0) associated with #36114.

## 🌍  Your Environment

**Angular Version:**
*Warning:* This is a build off of CI assets, so this isn't entirely accurate.

<pre><code>
Angular CLI: 9.0.6
Node: 10.14.2
OS: darwin x64

Angular: 9.1.0-next.4
... animations, common, compiler, compiler-cli, core, elements
... forms, language-service, platform-browser
... platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.6
@angular-devkit/build-angular     0.900.6
@angular-devkit/build-optimizer   0.900.6
@angular-devkit/build-webpack     0.900.6
@angular-devkit/core              9.0.6
@angular-devkit/schematics        9.0.6
@angular/cli                      9.0.6
@ngtools/webpack                  9.0.6
@schematics/angular               9.0.6
@schematics/update                0.900.6
rxjs                              6.5.4
typescript                        3.6.5
webpack                           4.41.2
</code></pre>

### PR DESCRIPTION
**Pull request by gkalpak**
~~_This PR sits on top of #36114 (to avoid conflicts), so only the last two commits are new._~~

##
Previously, event listeners for component output events attached on an Angular custom element before inserting it into the DOM (i.e. before instantiating the underlying component) didn't fire for events emitted during initialization lifecycle hooks, such as `ngAfterContentInit`, `ngAfterViewInit`, `ngOnChanges` (initial call) and `ngOnInit`.
The reason was that that `NgElementImpl` [subscribed to events][1] _after_ calling [ngElementStrategy#connect()][2], which is where the [initial change detection][3] takes place (running the initialization lifecycle hooks).

This commit fixes this by:
1. Ensuring `ComponentNgElementStrategy#events` is defined and available for subscribing to, even before instantiating the component.
2. Ensuring `NgElementImpl` subscribes to `NgElementStrategy#events` before calling `NgElementStrategy#connect()` (which initializes the component instance).

Jira issue: [FW-2010](https://angular-team.atlassian.net/browse/FW-2010)

Fixes #36141.

##
For future reference:
This PR increases the main bundle by ~190B (currently 453658B --> 453847B). This is caused by the newly added `=== null`/`!== null` checks, the extra `this.componentRef`s passed as arguments to methods and the `eventEmitters` and `events` initialization 🤦‍♂ 

[1]: https://github.com/angular/angular/blob/c0143cb2abdd172de1b95fd1d2c4cfc738640e28/packages/elements/src/create-custom-element.ts#L167-L170
[2]: https://github.com/angular/angular/blob/c0143cb2abdd172de1b95fd1d2c4cfc738640e28/packages/elements/src/create-custom-element.ts#L164
[3]: https://github.com/angular/angular/blob/c0143cb2abdd172de1b95fd1d2c4cfc738640e28/packages/elements/src/component-factory-strategy.ts#L158

### CODE REVIEW

File: aio/scripts/_payload-limits.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
     "master": {
       "uncompressed": {
         "runtime-es2015": 2987,
-        "main-es2015": 451469,
+        "main-es2015": 452031,
```
Comment (gkalpak) [inline aio/scripts/_payload-limits.json]: This ~190B increase (for the commits that are new in this PR) is caused by the newly added `=== null`/`!== null` checks, the extra `this.componentRef`s passed as arguments to methods and the `eventEmitters` and `events` initialization 🤦‍♂

**Comment by gkalpak** [comment]
MArking this as blocked on #36114, because that had to go in first.

**Review by andrewseguin — APPROVED** [review]
LGTM

**Comment by gkalpak** [comment]
BTW, here is a green CI run with the same commits against 9.1.x: https://circleci.com/workflow-run/27efe29b-0ac1-4755-9518-db324a588708

**Comment by gkalpak** [comment]
This requires a payload size bump for ViewEngine on 9.1.x.
Let's merge this into master only. I have prepared #37231 for 9.1.x (patch).

**Pull request closed without merging**